### PR TITLE
Checking if technique is apt

### PR DESCRIPTION
**Pull request by armenbeck**

### CODE REVIEW

File: paddy/Default_Numerics.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+"""default functions for numeric problems
+"""
+
+from .data_manager import paddy_parameter
+
+def p_hp_maker(length,scope,gausian_type):
```
Comment (frodofine) [inline paddy/Default_Numerics.py]: This is not correct. Please read up on object constructors in Python.

File: paddy/Default_Numerics.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
+"""default functions for numeric problems
```
Comment (frodofine) [inline paddy/Default_Numerics.py]: My python is not the best, but I believe this needs to go after the definition of the function/class. Not before.

File: paddy/data_manager.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
+""" Functions and Classes that deal with the data generated using paddy
```
Comment (frodofine) [inline paddy/data_manager.py]: See above.

File: paddy/data_manager.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            return(b)
+
+def get_generation(paddy_data,paddy_counter):
+    return(paddy_data['it{0}'.format(paddy_counter)])
+
+def get_population(paddy_data):
+    return(paddy_data)
+
+#example paddy model class
+class mlp(object):
```
Comment (frodofine) [inline paddy/data_manager.py]: Why is this here? If this meant to be a subclass of paddy_parameter (**which it is not**), it typically will go in its own file.

File: paddy/muscle.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
+""" muscle for the actual paddy algorithm that the package is based on
```
Comment (frodofine) [inline paddy/muscle.py]: What does 'muscle' mean? If these are utility functions, you should call them utilities.

File: paddy/data_manager.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+Functions inclued: 
+get_generation - function that passes the data from a paddy generation
+get_population - function thawt passes the data from all seeds
+seed_bank - manages the actual seeds in a dictionary
+data_bank - manages data dictonary generated by seeds
+
+Classes included:
+
+parameter
```
Comment (frodofine) [inline paddy/data_manager.py]: Where? There is no `parameter` class here. Did you mean `paddy_parameter`?

File: paddy/data_manager.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+"""
+import numpy as np
+import random
+
+def norm_p(p_val):
```
Comment (frodofine) [inline paddy/data_manager.py]: This should be a method of `paddy_parameter`.

File: paddy/pfa.py [hunk 8]
```diff
@@ -0,0 +1,7 @@
+'''
+imports the functions together for clean syntax
```
Comment (frodofine) [inline paddy/pfa.py]: Usually goes in `__init__`.

File: paddy/tests/__init__.py [hunk 9]
```diff
@@ -0,0 +1,4 @@
+
```
Comment (frodofine) [inline paddy/tests/__init__.py]: This file is typically blank.

File: paddy/data_manager.py (hunk 6 above)
Comment (frodofine) [reply paddy/data_manager.py]: Also, this file should be called something like **paddy_parameter.py**, not **data_manager.py**

File: paddy/__init__.py [hunk 10]
```diff
@@ -1,5 +1,12 @@
 
 #import stuff here
 
-__version__ = '0.2'
 
+
+from .utils import paddy_parameter, paddy_parameter, paddy_recover, get_param_names, random_propogation, paddy_runner
+from .writer import *
+
+
```
Comment (frodofine) [inline paddy/__init__.py]: Why all of this blank space?

File: paddy/tests/__init__.py [hunk 11]
```diff
@@ -0,0 +1,4 @@
...
+#import stuff here
```
Comment (frodofine) [inline paddy/tests/__init__.py]: Is this comment needed?

File: paddy/tests/test_utils.py [hunk 12]
```diff
@@ -0,0 +1,164 @@
+import paddy
+from paddy.Default_Numerics import *
+import unittest
+import itertools
+import random
+
+
+#example paddy model class
+class mlp(object):
```
Comment (frodofine) [inline paddy/tests/test_utils.py]: Why does this inherit from `object`? Would it make more sense to have it inherit from a `paddy_problem` class which allows one to set up the problem to their liking (define parameters, objective function, etc)?

File: paddy/tests/test_utils.py [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+    def __init__(self):
+        self.dropout_1 = paddy_parameter(param_range=[0,0.5,0.05],param_type='continious',limits=[0,1], gausian='default', normalization = 'on')
```
Comment (frodofine) [inline paddy/tests/test_utils.py]: Instead of having one *master* `paddy_parameter` class, you should make subclasses like `continuous_paddy_parameter`(which is spelled wrong above) and `discrete_paddy_parameter`.

File: paddy/tests/test_utils.py [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+            if boolean == False:
+                self.assertEqual(ecludian_value,rand_val[0])
+            if boolean == True:
+                norm_rand=test_parameter.norm_p(rand_val[0])
+                self.assertEqual(ecludian_value,norm_rand)
+    
+class test_paddy_runner(unittest.TestCase):
+
+    def test_each_function_during_run(self):
+
```
Comment (frodofine) [inline paddy/tests/test_utils.py]: Maybe add a quick comment explaining what this test case is for?

File: paddy/tests/__init__.py [hunk 15]
```diff
@@ -0,0 +1,4 @@
...
+
+__version__ = '0.2'
```
Comment (frodofine) [inline paddy/tests/__init__.py]: Why export this variable twice? In `hyperopt` this entire file is blank.

File: paddy/__init__.py [hunk 16]
```diff
@@ -1,5 +1,12 @@
 
 #import stuff here
```
Comment (frodofine) [inline paddy/__init__.py]: Do you need this comment?

File: paddy/utils.py [hunk 17]
```diff
@@ -0,0 +1,539 @@
...
+        for i in np.arange(tg[0],tg[1]):
+            if self.seed_data['fitness']['seed_{0}'.format(i)] == max_fit:
+                #needs verbos handling bellow
+                if verbose == True: 
+                    print('seed_{0}'.format(i))
+                temp_names.append('seed_{0}'.format(i))
+        return(temp_names)
+
+
+def get_param_names(p_space):
```
Comment (frodofine) [inline paddy/utils.py]: Why not make the name of a parameter a property returned by `paddy_parameter`.

File: paddy/utils.py [hunk 18]
```diff
@@ -0,0 +1,539 @@
...
+            fit_list = self.generation_fitness[str(self.paddy_counter)]
+        if counter != False:
+            try:
+                counter = int(counter)
+                fit_list = self.generation_fitness[str(counter)]
+            except:
+                print('counter needs to be an integer or string of integer')
+                return
+        max_fit = max(fit_list)
+        if verbose==True:
```
Comment (frodofine) [inline paddy/utils.py]: if verbose

File: paddy/utils.py [hunk 19]
```diff
@@ -0,0 +1,539 @@
...
+            print(max_fit)
+        if fit_list.count(max_fit) > 1:
+            print('multiple seeds of maximum fitness value:'+str(max_fit))
+        temp_names = []
+        if counter == False:
```
Comment (frodofine) [inline paddy/utils.py]: if not counter

File: paddy/utils.py [hunk 20]
```diff
@@ -0,0 +1,539 @@
...
+            tg = self.generation_data[str(self.paddy_counter)]
+        if counter != False:
```
Comment (frodofine) [inline paddy/utils.py]: else

File: paddy/utils.py [hunk 21]
```diff
@@ -0,0 +1,539 @@
...
+        for i in self.seed_data['fitness']:
+            if self.seed_data['fitness'][i] == max_fit:
+                temp_names.append(i)
+        return(temp_names)
+
+    def get_generation_top_seed(self,counter=False,verbose=False):
+        ''' gets the max value of fitness in the generation
+        counter lets the user call the function after run to
+        analyze by hand'''
+        if counter == False:
```
Comment (frodofine) [inline paddy/utils.py]: not verbose

File: paddy/utils.py [hunk 22]
```diff
@@ -0,0 +1,539 @@
...
+            fit_list = self.generation_fitness[str(self.paddy_counter)]
+        if counter != False:
```
Comment (frodofine) [inline paddy/utils.py]: else

File: paddy/utils.py [hunk 23]
```diff
@@ -0,0 +1,539 @@
...
+        ##################################
+        print("paddy is done!")
+        if self.file_name != '':
+            self.save_paddy()
+
+    def paddy_plot_and_print(self, verbose = False):
+        '''writes to terminal results as specified and
+        plots paddy results as specified in 
+        plotter package under preformance_plotter '''
+        if verbose == False:
```
Comment (frodofine) [inline paddy/utils.py]: if not verbose

File: paddy/utils.py [hunk 24]
```diff
@@ -0,0 +1,539 @@
...
+            print("plot and print requires a string argument or list containing valid string arguments")
+            return
+        if verbose == True:
```
Comment (frodofine) [inline paddy/utils.py]: else

File: paddy/utils.py [hunk 25]
```diff
@@ -0,0 +1,539 @@
...
+            for comand in verbose:
+                if paddy_plot_and_write_comands.issuperset({comand}):
+                    pass
+                else:
+                    print(str(comand)+' is not a valid input for plot and print')
+                    verbose = False
+                    return
+            verbose = set(verbose)
+            #print(any( print_type == verbose for print_type in paddy_plot_and_write_comands.intersection(verbose) ))
+            if verbose != False:
```
Comment (frodofine) [inline paddy/utils.py]: What?

File: paddy/utils.py [hunk 26]
```diff
@@ -0,0 +1,539 @@
...
+        else:
+            print("recovered paddy run was already completed, instead of recovering run 'extend_paddy' to continue optimization with the saved paddy runner")
+    
+    
+    def extend_paddy(self, new_iterations , new_file_name= False , verbose=False):
+        'callable that extends the iterations of a completed paddy runner'
+        assert( isinstance(new_iterations, int) == True, "partial paddy runs are not a feature, use integers when extending the itteration parameter on a recoverd paddy runner")
+        assert(new_iterations > 0, "A negative input for new iterations terminates the run, value must be positive to extend parameters")
+        self.recover = True
+        if verbose != False:
```
Comment (frodofine) [inline paddy/utils.py]: if verbose

File: paddy/utils.py [hunk 27]
```diff
@@ -0,0 +1,539 @@
...
+            self.new_propogation()
+            if self.file_name != '':
+                self.save_paddy()
+        self.top_values[str(self.paddy_counter)]={}
+        seed_key_numbers = self.generation_data[str(self.paddy_counter)]
+        temp_gen_fit = []
+        for i in range(seed_key_numbers[0],seed_key_numbers[1]+1):
+            temp_gen_fit.append(self.seed_data['fitness']['seed_{0}'.format(i)])
+        self.generation_fitness[str(self.paddy_counter)]=temp_gen_fit
+        if self.paddy_type == 'generational':
```
Comment (frodofine) [inline paddy/utils.py]: There's got to be a better way to do this other than a bunch of `if` statements...

File: paddy/utils.py [hunk 28]
```diff
@@ -0,0 +1,539 @@
...
+            if self.yt > seed_key_numbers[1] - seed_key_numbers[0]:
+                self.yt = int(round((seed_key_numbers[1] - seed_key_numbers[0])*.75))#this adjusts the threshold operator in the case that it is set to be greater than seeds generated
+                #print(int(round((seed_key_numbers[1] - seed_key_numbers[0])*.75)))
+            gen_clone = {}
+            for i in range(seed_key_numbers[0],seed_key_numbers[1]+1):
+                gen_clone['seed_{0}'.format(i)]=self.seed_data['fitness']['seed_{0}'.format(i)]
+            gen_fitness_keys = sorted( gen_clone, key=lambda key:gen_clone[key])#list of seed number id's that are to be sown
+            y_max = self.seed_data['fitness'][gen_fitness_keys[-1]]
+            self.top_values[str(self.paddy_counter)]={}
```
Comment (frodofine) [inline paddy/utils.py]: Why are you using a string to store these in a dictionary when you could use an array instead?

File: paddy/utils.py [hunk 29]
```diff
@@ -0,0 +1,539 @@
...
+                #this will let the paddy run even if there are no neighbors
+                for i in n_values2:
+                    neighbors[i] = 1
+                print('you might want to tweek your paddy parameters, new seeds did not have neighbors')
+                break
+            for i in n_values2:
+                n_count = 0
+                for j in n_values2:
+                    if i != j:
+                        if (distance.euclidean(n_values2[i],n_values2[j])-np.quantile(d_list,quantile_value) < 0):
```
Comment (frodofine) [inline paddy/utils.py]: Use a class instead to allow for other distance metrics.

File: paddy/writer.py [hunk 30]
```diff
@@ -0,0 +1,119 @@
+try:
```
Comment (frodofine) [inline paddy/writer.py]: Use `pip` to handle requirements instead.

File: paddy/utils.py [hunk 31]
```diff
@@ -0,0 +1,539 @@
...
+                temp_list.append(n_values[key][i])
+            n_values2[key]= temp_list
+        neighbors = {}
+        d_list = []
+        for i in n_values2:
+            n_count = 0
+            for j in n_values2:
+                if i != j:
+                    d_list.append(distance.euclidean(n_values2[i],n_values2[j]))
+                    if (distance.euclidean(n_values2[i],n_values2[j])-self.r < 0):
```
Comment (frodofine) [inline paddy/utils.py]: What if I don't want to use Euclidean distance want to use, say Tanimoto distance?

File: paddy/utils.py [hunk 32]
```diff
@@ -0,0 +1,539 @@
...
+                    if i != j:
+                        if (distance.euclidean(n_values2[i],n_values2[j])-np.quantile(d_list,quantile_value) < 0):
+                            n_count += 1
+                neighbors[i] = n_count
+            quantile_value -= 0.05
+        n_max=(max(list(neighbors.values())))
+        self.Un = {}
+        for i in n_values2:
+            self.Un[i]= math.exp((neighbors[i]/n_max)-1)
+        self.S = {} # dictionary of polinated seeds Un  * s (self.s)
```
Comment (frodofine) [inline paddy/utils.py]: Use arrays instead!

File: paddy/utils.py [hunk 33]
```diff
@@ -0,0 +1,539 @@
...
+        for i in n_values2:
+            n_count = 0
+            for j in n_values2:
+                if i != j:
+                    d_list.append(distance.euclidean(n_values2[i],n_values2[j]))
+                    if (distance.euclidean(n_values2[i],n_values2[j])-self.r < 0):
+                        n_count += 1
+            neighbors[i] = n_count
+        quantile_value = 0.75
+        while all( x == 0 for x in neighbors.values()):
```
Comment (frodofine) [inline paddy/utils.py]: Why is this a loop?

File: paddy/Default_Numerics.py [hunk 34]
```diff
@@ -0,0 +1,104 @@
...
+def eval_gl(seed):
+    '''default function for evaluating fitting to gramacy lee '''
+    x,y = gramacy_lee()
+    answer = trig_inter(x,seed,dim=len(seed))
+    #made negative for maximization problem
+    error = -fitting_error_function(y,answer)
+    return error  
+
+
+class polynomial(object):
```
Comment (frodofine) [inline paddy/Default_Numerics.py]: Should this derive from a class like `paddy_space`?

File: paddy/utils.py [hunk 35]
```diff
@@ -0,0 +1,539 @@
...
+            print('multiple seeds of maximum fitness value:'+str(max_fit))
+        temp_names = []
+        if counter == False:
+            tg = self.generation_data[str(self.paddy_counter)]
+        if counter != False:
+            tg = self.generation_data[str(counter)]
+        for i in np.arange(tg[0],tg[1]):
+            if self.seed_data['fitness']['seed_{0}'.format(i)] == max_fit:
+                #needs verbos handling bellow
+                if verbose == True:
```
Comment (frodofine) [inline paddy/utils.py]: if verbose

File: paddy/utils.py [hunk 36]
```diff
@@ -0,0 +1,539 @@
...
+            print("plot and print requires a string argument or list containing valid string arguments")
+            return
+        if verbose == True:
+            print('plot and print does not accept bool objects')
+            return
+        print(verbose)
+        paddy_plot_and_write_comands = {'generation', 'final_results', 'pop_fitness',
+        'gen_fitness','best_sown','average_population', "scatter",'box','box_notched','box_hidden','average_gen'}
+        plotter_types = {'best_sown','average_population','scatter','box','box_notched','box_hidden','average_gen'}
+        if verbose != False:
```
Comment (frodofine) [inline paddy/utils.py]: if not verbose?

The rest of this statement doesn't make sense to me, why are you checking if its a string?

File: paddy/utils.py [hunk 37]
```diff
@@ -0,0 +1,539 @@
...
+           
+           the second half generates the Un variable, and calculates a new r if no neigbors are present'''
+        p_names = get_param_names(self.space) # allows us to append ecludian values to list without loosing order
+        n_values = {} # dictionary of seed{param:val,param2:val}...
+        for key in self.s:
+            n_values[key] = {}
+            for parameter in self.seed_data['parameters'][key]:
+                temp = getattr(self.space,parameter).get_ecludian_values(self.seed_data['parameters'][key][parameter]) #this line generates the value for ecludian
+                n_values[key][parameter] = temp
+        n_values2 = {} # dictionary of seed:[val1,val2]
```
Comment (frodofine) [inline paddy/utils.py]: Array instead maybe?

File: paddy/utils.py (hunk 33 above)
Comment (armenbeck) [reply paddy/utils.py]: To prevent passing neighbor values of only 0, it would make the rest crash

File: paddy/utils.py (hunk 36 above)
Comment (armenbeck) [reply paddy/utils.py]: strings don't allow for subset checking where sets do, so a single string is placed into a set.

File: paddy/utils.py [hunk 38]
```diff
@@ -0,0 +1,539 @@
...
+                                b[1] = np.random.normal(loc=values[1],scale = 0.2)
+                            is_done = True
+            if self.normalization == True:
+                b_norm = [np.random.normal(loc= self.norm_p(values[0]) , scale = 0.2 **(10**values[1]))]
+                g = ( b_norm[0] * (self.limits[1]-self.limits[0]) ) + self.limits[0]
+                if g > self.limits[0]:
+                    if g < self.limits[1]:
+                        b = [g]
+                        b.append(0)
+                        if self.gausian == 'scaled':
```
[the remark this answers is not in the document]
Comment (armenbeck) [reply paddy/utils.py]: It is left as a string in case other string arguments are added by users or myself in a patch

File: paddy/utils.py [hunk 39]
```diff
@@ -0,0 +1,539 @@
...
+        if self.param_type == 'integer':
+           value[0] = int(round(value[0]))
+        value.append(0)
+        return(value)
+    def new_seed_init(self, values):
+        '''takes values[param_value][gausian_value] returns new value(s)
+            does this while considering normalization and gausian type'''
+        is_done = False
+        b = []
+        while not is_done:
```
[the remark this answers is not in the document]
Comment (armenbeck) [reply paddy/utils.py]: No, it repeats if a value does not fit to the limit bounds

File: paddy/utils.py [hunk 40]
```diff
@@ -0,0 +1,539 @@
...
+
+
+class paddy_parameter(object):
+    def __init__(self, param_range, param_type, limits, gausian, normalization):
+        self.param_range = param_range
+        self.param_type = param_type
+        self.limits = limits
+        self.gausian = gausian
+        self.normalization = normalization
+        #need to assert that limits and param_range match
```
[the remark this answers is not in the document]
Comment (armenbeck) [reply paddy/utils.py]: If you initialize with a random value outside of the limits, it might never produce a value that satisfies the limits term in an endless loop

File: paddy/utils.py [hunk 41]
```diff
@@ -0,0 +1,539 @@
...
+        return(value)
+    def new_seed_init(self, values):
+        '''takes values[param_value][gausian_value] returns new value(s)
+            does this while considering normalization and gausian type'''
+        is_done = False
+        b = []
+        while not is_done:
+            if self.normalization != True:
+                b = np.random.normal(loc= values[0] , scale = 0.2 **(10**values[1]))
+                if self.limits == False:
```
[the remark this answers is not in the document]
Comment (armenbeck) [reply paddy/utils.py]: changed limits to None, and set line 47 to 'if not self.normalization:'

File: paddy/utils.py (hunk 38 above)
[the remark this answers is not in the document]
Comment (frodofine) [reply paddy/utils.py]: This is not a good strategy. Think about creating a Gaussian class that does this for you.

File: paddy/utils.py (hunk 39 above)
[the remark this answers is not in the document]
Comment (frodofine) [reply paddy/utils.py]: Why don't you use clamping instead?

File: paddy/utils.py [hunk 42]
```diff
@@ -0,0 +1,539 @@
...
+                    b = [b,0]
+                    if self.gausian == 'scaled':
+                        b[1] = np.random.normal(loc=values[1],scale = 0.2)
+                    is_done = True
+                if self.limits != False:
```
[the remark this answers is not in the document]
Comment (armenbeck) [reply paddy/utils.py]: changed line 54 to 'elif self.limits !- None:'

File: paddy/utils.py (hunk 40 above)
[the remark this answers is not in the document]
Comment (frodofine) [reply paddy/utils.py]: No, why are you asserting that if `limits == False` then `normalization == False`? What does that assertion have to do with `param_range`?

File: paddy/utils.py (hunk 38 above)
[the remark this answers is not in the document]
Comment (armenbeck) [reply paddy/utils.py]: Sounds good, I think it will be easy to pull the Gaussian type value when set as a class and have the original code remain mostly the same

File: paddy/utils.py (hunk 40 above)
[the remark this answers is not in the document]
Comment (armenbeck) [reply paddy/utils.py]: *Resolved via dialogue, may need to use class vs flagging

File: paddy/utils.py [hunk 43]
```diff
@@ -0,0 +1,539 @@
...
+                    if b > self.limits[0]:
+                        if b < self.limits[1]:
+                            b = [b,0]
+                            if self.gausian == 'scaled':
+                                b[1] = np.random.normal(loc=values[1],scale = 0.2)
+                            is_done = True
+            if self.normalization == True:
```
[the remark this answers is not in the document]
Comment (armenbeck) [reply paddy/utils.py]: changed to 'else:'

File: paddy/utils.py [hunk 44]
```diff
@@ -0,0 +1,539 @@
...
+    def new_seed_init(self, values):
+        '''takes values[param_value][gausian_value] returns new value(s)
+            does this while considering normalization and gausian type'''
+        is_done = False
+        b = []
+        while not is_done:
+            if self.normalization != True:
+                b = np.random.normal(loc= values[0] , scale = 0.2 **(10**values[1]))
+                if self.limits == False:
+                    b = [b,0]
```
[the remark this answers is not in the document]
Comment (armenbeck) [reply paddy/utils.py]: see changes above
